nep-sales.mediawiki

```diff
@@ -0,0 +1,123 @@
+
+2. The investor depends on an announcement for start and closing of the sales from the organisation. However, many people still do not do their research and proceed to participate in a closed sale, causing lost funds and anger.
+
+3. There might be contribution limits that are recorded on the blockchain but it is opaque to the wallet software. Thus, the wallet is unable to stop the investor from over-contributing.
+
+4. There is currently no records of the entire sales process. Tokens can be created at different rates at different times. It is confusing to tell what is the current exchange rate without attempting a transaction first.
+
+This proposal tackles the mentioned issues by proposing a standard set of methods that wallets can implement to provide investors with the visibility and confidence when contributing. With the proposed set of methods, the wallet can independently verify the sale and verification status, providing the investor with confidence in the sales.
+
+It is noted that the above issues are all tackled by using verification and having the contract reject transactions. However, this is a very unfriendly way of rejection with no clarity of the reason of rejection given to the investor. With this proposal, we hope that the token sales experience will be smooth and painless. This proposal is not an excuse to not implement verification as it is still the better option to implement.
+
+==Specification==
+
+We propose the following methods to be implemented in any crowdsale contract that requires investor verification as well as sales restricted to a timeframe.
+
+The methods starting with "kyc" are methods for individuals, indicating their status in the current sale.
+
+The methods starting with "sale" are methods for the entire sale, indicating the status of the current sale.
+
+This standard defines two method types:
+
+* '''(Required)''' : methods that are compulsory in order to be NEP-compliant. Methods are required unless marked otherwise. These methods are mostly user-facing and thus must be standardised for a smooth experience.
+
+* '''(Optional)''' : methods that are optionally implemented. These methods are not all user-facing and contains certains parameters not present in all sales. However, developer tools are encouraged to follow this specification for the best experience.
+
+===Methods===
+
+====mintTokens====
+
+* Syntax: <code>public static string mintTokens(byte[] address)</code>
+
+* Remarks: ''mintToken'' attempts to create tokens for the specified address. If successful, this action should increase the totalSupply of the tokens. This method must be implemented together with the ''token_mint'' event.
+
+====kycStatus====
+
+* Syntax: <code>public static bool kycStatus(byte[] address)</code>
+
+* Remarks: ''kycStatus'' Checks the whitelist status for the provided address. Returns True if registered.
+
+====kycLimit====

```

Comment:
While this is a useful method in some specific sales cases, it might not be appropriate as a standard method. 

- Depending on the token sale, there are different limits; with some sales having multiple limited rounds. 
- Doesn't fit if there are sales that have no limits, and returning the `largest possible integer` seems a bit hacky.
- Doesn't address sales where there are minimum contribution requirements.

Side note: if this method were to be kept, a different name may be more appropriate. `kycLimit` may seems to suggest there is a limit pertaining to the kyc process or number of kyc addresses. This is really more a part of the sales methods.

It may just be better to remove this method, and have contract owners provide users with information about round limits as a part of `saleDetails`. `saleDetails` can also be updated to take an optional address parameter to allow contracts to provide user specific details about their contributions to date.

Reply:
i will remove this in favor of making `saleDetails` be the main source of information since it will be difficult in ensuring this information stays relevant for all token sales.